### PR TITLE
Python interface for MatmulParams and example script

### PR DESCRIPTION
**Pull request by jacobhinkle**
Implements the Python interface for `MatmulParams`, so that we can manually set parameters and schedule matmul fusions from the Python frontend. Included is a simple script demonstrating how to do a parameter sweep and measure perf for a couple parameters.

### CODE REVIEW

File: csrc/python_frontend/python_bindings.cpp [hunk 1]
```diff
@@ -677,48 +677,65 @@ void initNvFuserPythonBindings(PyObject* module) {
   heuristic_params.def(
       "__repr__", [](const HeuristicParams& self) { return self.toString(); });
 
+#define INITPARAMS(cfg, internal_type)                                     \
+  py::class_<internal_type, HeuristicParams> cfg(nvfuser, #internal_type); \
+  cfg.def(py::init());                                                     \
+  cfg.def(                                                                 \
+      "__repr__", [](const internal_type& self) { return self.toString(); });
+#define PARAM(cfg, internal_type, type, name)        \
```
Comment (jacobhinkle) [inline csrc/python_frontend/python_bindings.cpp:685]: TODO: we should probably add a `.doc("foo")` to these params and to the "_config" type to make them more user-friendly

**Comment by rdspring1** [comment]
https://nvidia.slack.com/archives/D03NWK6FMLG/p1729496574125999

How to get nvfuser `FusionDefinition` from Thunder trace? Now you can just write a single pytorch eager definition and get the corresponding nvfuser definition?

File: csrc/python_frontend/python_bindings.cpp [hunk 2]
```diff
@@ -677,48 +677,75 @@ void initNvFuserPythonBindings(PyObject* module) {
...
+      [](MatmulParams& self, MatmulParams::TileRasterizationOrder cta_order_) {
+        self.cta_order = cta_order_;
       });
-  pointwise_config.def_property(
-      "unroll_factor",
-      [](PointwiseParams& self) { return self.unroll_factor; },
-      [](PointwiseParams& self, int64_t unroll_factor_) {
-        self.unroll_factor = unroll_factor_;
-      });
-  pointwise_config.def(
```
Comment (rdspring1) [inline csrc/python_frontend/python_bindings.cpp:720]: Does `__repr__` still exist? It would be useful for matmul params too.

Comment (jacobhinkle) [reply csrc/python_frontend/python_bindings.cpp:720]: Yeah, it moved to INITPARAMS macro above. I used it for debugging my script. Works as expected: `print(params)`

Comment (rdspring1) [reply csrc/python_frontend/python_bindings.cpp:720]: Ah, I'm blind. 🙈

File: csrc/python_frontend/python_bindings.cpp [hunk 3]
```diff
@@ -677,48 +678,91 @@ void initNvFuserPythonBindings(PyObject* module) {
   heuristic_params.def(
       "__repr__", [](const HeuristicParams& self) { return self.toString(); });
 
+#define INITPARAMS(cfg, internal_type)                                     \
```
Comment (rdspring1) [inline csrc/python_frontend/python_bindings.cpp:681]: Should we define all macros upfront?
1. INITPARAMS
2. PARAM
3. TILEPARAM
4. CBPARAM

File: csrc/python_frontend/python_bindings.cpp [hunk 4]
```diff
@@ -677,48 +678,91 @@ void initNvFuserPythonBindings(PyObject* module) {
...
+  mma_macro.def_property_readonly(
+      "m", [](MmaMacro& self) -> uint16_t { return MmaMacroEncode(self).m; });
+  mma_macro.def_property_readonly(
+      "n", [](MmaMacro& self) -> uint16_t { return MmaMacroEncode(self).n; });
+  mma_macro.def_property_readonly(
+      "k", [](MmaMacro& self) -> uint16_t { return MmaMacroEncode(self).k; });
+  mma_macro.def("__repr__", [](MmaMacro& self) { return toString(self); });
+  PARAM(matmul_config, MatmulParams, MmaMacroEncode, mma_macro);
+
+#undef PARAM
```
Comment (rdspring1) [inline csrc/python_frontend/python_bindings.cpp:764]: Then, undefine all macros here.

**Review by rdspring1 — APPROVED** [review]
LGTM.

How about defining all macros upfront and then deleting them after defining all parameter classes? Then, you can read the parameter classes in a contiguous block.

File: csrc/python_frontend/python_bindings.cpp (hunk 3 above)
Comment (jacobhinkle) [reply csrc/python_frontend/python_bindings.cpp:681]: `TILEPARAM` and `CBPARAM` are specific to matmul, so I now define those at the top of the MatmulParams section, as well as the `py::class_` and `py::enum_` definitions. This way you can easily read out the `matmul_config` definition.

File: version.txt [hunk 5]
```diff
@@ -1 +1 @@
-0.2.18
+0.2.19
```
Comment (rdspring1) [inline version.txt:1]: The current version is `0.2.20`. Probably should bump this `0.2.21`.

**Comment by jacobhinkle** [comment]
!build

**Comment by jacobhinkle** [comment]
!build

File: csrc/scheduler/matmul_heuristic.h [hunk 6]
```diff
@@ -201,7 +201,7 @@ class MatmulParams : public HeuristicParams {
                                                                   : "int32_t")
                : "unavailable")
        << "\n"
-       << "Tile rastrization order: "
+       << "Tile rasterization order: "
```
Comment (jacobhinkle) [inline csrc/scheduler/matmul_heuristic.h:204]: Just fixing a typo

**Pull request merged**